service-wfs-client/src/main/java/org/oskari/service/wfs/client/OskariWFSFilter.java

```diff
@@ -0,0 +1,132 @@
+                CRS.toSRS(bbox.getCoordinateReferenceSystem()));
+    }
+
+    protected static Filter getFilter(JSONObject filterJson) {
+        Filter filter = null;
+        JSONObject prop = JSONHelper.getJSONObject(filterJson, KEY_PROPERTY);
+        if (prop != null) {
+            filter = getPropertyFilter(prop);
+        }
+        List <Filter> ands = JSONHelper.getArrayAsList(JSONHelper.getJSONArray(filterJson, KEY_AND))
+                .stream().map(f -> getPropertyFilter((JSONObject) f)).collect(Collectors.toList());
+        if (!ands.isEmpty()) {
+            filter = appendFilter(filter, ff.and(ands));
+        }
+        List <Filter> ors = JSONHelper.getArrayAsList(JSONHelper.getJSONArray(filterJson, KEY_OR))
+                .stream().map(f -> getPropertyFilter((JSONObject) f)).collect(Collectors.toList());
+        if (!ors.isEmpty()) {
+            filter = appendFilter(filter, ff.or(ors));
+        }
+
+        return filter;
+    }
+    private static Filter appendFilter(final Filter main, final Filter toAppend) {
+        if (main == null) {
+            return toAppend;
+        } else {
+            return ff.and(main, toAppend);
+        }
+    }
+    private static Filter getPropertyFilter(JSONObject property) {
+        String key = JSONHelper.optString(property, KEY);
+        boolean match = JSONHelper.getBooleanFromJSON(property,KEY_MATCH_CASE, true);
+        Expression name = ff.property(key);
+        if (property.has(KEY_EQUAL)) {
+            return ff.equal(name, ff.literal(JSONHelper.get(property, KEY_EQUAL)), match);
+        }
+        if (property.has(KEY_GREATER)) {
+            Filter greater = ff.greater(name, ff.literal(JSONHelper.get(property, KEY_GREATER)));
+            if (property.has(KEY_LESS)) {
+                Filter less = ff.less(name, ff.literal(JSONHelper.get(property, KEY_LESS)));
+                return ff.and(greater, less);
+            }
+            if (property.has(KEY_LESS_OR_EQUAL)) {
+                Filter lessEqual = ff.lessOrEqual(name, ff.literal(JSONHelper.get(property, KEY_LESS_OR_EQUAL)));
+                return ff.and(greater, lessEqual);
+            }
+            return greater;
+        }
+        if (property.has(KEY_IN)) {
+            List <Filter> values = JSONHelper.getArrayAsList(JSONHelper.getJSONArray(property,KEY_IN))
+                    .stream()
+                    .map(value -> ff.equal(name, ff.literal(value), match))
+                    .collect(Collectors.toList());
+            return ff.or(values);
+        }
+        if (property.has(KEY_NOT_IN)) {
+            List <Filter> values = JSONHelper.getArrayAsList(JSONHelper.getJSONArray(property,KEY_NOT_IN))
+                    .stream()
+                    .map(value -> ff.notEqual(name, ff.literal(value), match))
+                    // .map(filter -> property.has(KEY_NOT_IN) ? ff.not(filter):filter)
+                    .collect(Collectors.toList());
+            return ff.or(values);
```

Comment:
`NOT IN (A, B) <=> NOT (A OR B) <=> NOT A AND NOT B` is not equivalent to `NOT A OR NOT B`!